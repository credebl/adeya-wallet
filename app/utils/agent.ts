import { Agent, getAgentModules, useAdeyaAgent } from '@adeya/ssi'

export type AdeyaAgent = Agent<ReturnType<typeof getAgentModules>>
export type AdeyaAgentModules = ReturnType<typeof getAgentModules>

interface MyAgentContextInterface {
  loading: boolean
  agent: AdeyaAgent
  setAgent: (agent: AdeyaAgent) => void
}

export const useAppAgent = useAdeyaAgent as () => MyAgentContextInterface
