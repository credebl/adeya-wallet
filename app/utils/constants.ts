export const hitSlop = { top: 44, bottom: 44, left: 44, right: 44 }
