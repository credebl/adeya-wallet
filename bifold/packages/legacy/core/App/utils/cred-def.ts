import { CredentialMetadataKeys, CredentialExchangeRecord as CredentialRecord } from '@aries-framework/core'

import { parseSchemaFromId, credentialSchema } from './schema'

export function parseCredDefFromId(credDefId?: string, schemaId?: string): string {
  let name = 'Credential'
  if (credDefId) {
    const credDefRegex = /[^:]+/g
    const credDefIdParts = credDefId.match(credDefRegex)
    if (credDefIdParts?.length === 5) {
      name = `${credDefIdParts?.[4].replace(/_|-/g, ' ')}`
        .split(' ')
        .map((credIdPart) => credIdPart.charAt(0).toUpperCase() + credIdPart.substring(1))
        .join(' ')
    }
  }
  if (name.toLocaleLowerCase() === 'default' || name.toLowerCase() === 'credential') {
    name = parseSchemaFromId(schemaId).name
  }
  return name
}

function credentialDefinition(credential: CredentialRecord): string | undefined {
  return credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.credentialDefinitionId
}

export function parsedCredDefNameFromCredential(credential: CredentialRecord): string {
  return parseCredDefFromId(credentialDefinition(credential), credentialSchema(credential))
}

export function parsedCredDefName(credentialDefinitionId: string, schemaId: string): string {
  return parseCredDefFromId(credentialDefinitionId, schemaId)
}
