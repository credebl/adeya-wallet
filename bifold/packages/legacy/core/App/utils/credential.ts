import { CredentialExchangeRecord, CredentialMetadataKeys, CredentialState } from '@aries-framework/core'
import { ImageSourcePropType } from 'react-native'

import { luminanceForHexColor } from './luminance'

export const isValidIndyCredential = (credential: CredentialExchangeRecord) => {
  return (
    credential &&
    (credential.state === CredentialState.OfferReceived ||
      credential.credentials.find((c) => c.credentialRecordType === 'indy'))
  )
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const credentialTextColor = (ColorPallet: any, hex?: string) => {
  const midpoint = 255 / 2
  if ((luminanceForHexColor(hex ?? '') ?? 0) >= midpoint) {
    return ColorPallet.grayscale.darkGrey
  }
  return ColorPallet.grayscale.white
}

export const toImageSource = (source: unknown): ImageSourcePropType => {
  if (typeof source === 'string') {
    return { uri: source as string }
  }
  return source as ImageSourcePropType
}

export const getCredentialIdentifiers = (credential: CredentialExchangeRecord) => {
  return {
    credentialDefinitionId: credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.credentialDefinitionId,
    schemaId: credential.metadata.get(CredentialMetadataKeys.IndyCredential)?.schemaId,
  }
}
