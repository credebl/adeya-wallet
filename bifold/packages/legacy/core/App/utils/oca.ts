import { CredentialExchangeRecord } from '@aries-framework/core'

import { CredentialSharedProofData, IndyProofRequestTemplatePayloadData } from '../../verifier'
import { Attribute, Field, Predicate } from '../types/record'

export const buildFieldsFromIndyCredential = (credential: CredentialExchangeRecord): Array<Field> => {
  return credential?.credentialAttributes?.map((attr) => new Attribute(attr)) || []
}

export const buildFieldsFromIndyProofRequestTemplate = (data: IndyProofRequestTemplatePayloadData): Array<Field> => {
  const fields = []
  if (data.requestedAttributes) {
    for (const item of data.requestedAttributes) {
      if (item.name) {
        fields.push(new Attribute({ name: item.name, value: null, ...item }))
      }
      if (item.names) {
        for (const name of item.names) {
          fields.push(new Attribute({ name, value: null, ...item }))
        }
      }
    }
  }
  if (data.requestedPredicates) {
    for (const item of data.requestedPredicates) {
      fields.push(
        new Predicate({
          pType: item.predicateType,
          pValue: item.predicateValue,
          ...item,
        })
      )
    }
  }
  return fields
}

export const buildFieldsFromSharedIndyProof = (data: CredentialSharedProofData): Array<Field> => {
  const fields = []
  for (const attribute of data.sharedAttributes) {
    fields.push(
      new Attribute({
        name: attribute.name,
        value: attribute.value,
      })
    )
  }
  for (const attributesGroup of data.sharedAttributeGroups) {
    for (const attribute of attributesGroup.attributes) {
      fields.push(
        new Attribute({
          name: attribute.name,
          value: attribute.value,
        })
      )
    }
  }
  for (const predicate of data.resolvedPredicates) {
    fields.push(
      new Predicate({
        name: predicate.name,
        pType: predicate.predicateType,
        pValue: predicate.predicateValue,
      })
    )
  }
  return fields
}
