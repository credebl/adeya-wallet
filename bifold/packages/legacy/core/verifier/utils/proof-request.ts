import {
  Agent,
  AgentMessage,
  AttributeFilter,
  AttributeValue,
  AutoAcceptProof,
  ProofAttributeInfo,
  ProofExchangeRecord,
  ProofPredicateInfo,
  ProofRequest,
  V1RequestPresentationMessage,
} from '@aries-framework/core'

import { ProofRequestTemplate, ProofRequestType } from '../types/proof-reqeust-template'

const protocolVersion = 'v1'
const domain = 'http://aries-mobile-agent.com'

/*
 * Find Proof Request message in the storage by the given id
 * */
export const findProofRequestMessage = async (agent: Agent, id: string): Promise<ProofRequest | undefined> => {
  const message = await agent.proofs.findRequestMessage(id)
  if (message && message instanceof V1RequestPresentationMessage && message.indyProofRequest) {
    return message.indyProofRequest
  } else {
    return undefined
  }
}

const convertCase: { [key: string]: keyof AttributeFilter | undefined } = {
  schema_id: 'schemaId',
  schema_issuer_did: 'schemaIssuerDid',
  schema_name: 'schemaName',
  schema_version: 'schemaVersion',
  issuer_did: 'issuerDid',
  cred_def_id: 'credentialDefinitionId',
}

const fromAnonCreds = (restriction: Record<string, string>): AttributeFilter => {
  const restrictionObject = {} as AttributeFilter
  Object.keys(restriction).forEach((anonKey) => {
    if (anonKey.startsWith('attr::') && anonKey.endsWith('::value')) {
      restrictionObject.attributeValue = new AttributeValue({
        name: anonKey.split('::')[1],
        value: restriction[anonKey],
      })
    } else {
      const key = convertCase[anonKey]
      if (key) {
        restrictionObject[key] = restriction[anonKey] as any
      }
    }
  })
  return restrictionObject
}

/*
 * Build Proof Request data from for provided template
 * */
/*
 * Build Proof Request data for provided template
 * */
export const buildProofRequestDataForTemplate = (
  template: ProofRequestTemplate,
  customValues?: Record<string, Record<string, number>>
) => {
  if (template.payload.type === ProofRequestType.Indy) {
    const requestedAttributes: Map<string, ProofAttributeInfo> = new Map()
    const requestedPredicates: Map<string, ProofPredicateInfo> = new Map()
    let index = 0

    template.payload.data.forEach((data) => {
      if (data.requestedAttributes?.length) {
        data.requestedAttributes.forEach((requestedAttribute) => {
          const attribute = new ProofAttributeInfo({
            name: requestedAttribute.name,
            names: requestedAttribute.names,
            nonRevoked: requestedAttribute.nonRevoked,
            restrictions: requestedAttribute.restrictions?.map((restriction) => fromAnonCreds(restriction)),
          })
          requestedAttributes.set(`referent_${index}`, attribute)
          index++
        })
      }
      if (data.requestedPredicates?.length) {
        data.requestedPredicates.forEach((requestedPredicate) => {
          const customValue =
            customValues && customValues[data.schema] ? customValues[data.schema][requestedPredicate.name] : undefined

          const predicate = new ProofPredicateInfo({
            name: requestedPredicate.name,
            predicateValue:
              requestedPredicate.parameterizable && customValue ? customValue : requestedPredicate.predicateValue,
            predicateType: requestedPredicate.predicateType,
            nonRevoked: requestedPredicate.nonRevoked,
            restrictions: requestedPredicate.restrictions?.map((restriction) => fromAnonCreds(restriction)),
          })

          requestedPredicates.set(`referent_${index}`, predicate)
          index++
        })
      }
    })
    return {
      indy: {
        name: template.name,
        version: template.version,
        nonce: Date.now().toString(),
        requestedAttributes,
        requestedPredicates,
      },
    }
  }
  if (template.payload.type === ProofRequestType.DIF) {
    return {}
  }
}

export interface CreateProofRequestInvitationResult {
  request: AgentMessage
  proofRecord: ProofExchangeRecord
  invitation: AgentMessage
  invitationUrl: string
}

/*
 * Create connectionless proof request invitation for provided template
 * */
export const createConnectionlessProofRequestInvitation = async (
  agent: Agent,
  template: ProofRequestTemplate,
  customPredicateValues?: Record<string, Record<string, number>>
): Promise<CreateProofRequestInvitationResult | undefined> => {
  const proofFormats = buildProofRequestDataForTemplate(template, customPredicateValues)
  if (!proofFormats) {
    return undefined
  }
  const { message: request, proofRecord } = await agent.proofs.createRequest({
    protocolVersion,
    autoAcceptProof: AutoAcceptProof.Always,
    proofFormats,
  })
  const { message: invitation, invitationUrl } = await agent.oob.createLegacyConnectionlessInvitation({
    recordId: proofRecord.id,
    message: request,
    domain,
  })
  return {
    request,
    proofRecord,
    invitation,
    invitationUrl,
  }
}

export interface SendProofRequestResult {
  proofRecord: ProofExchangeRecord
}

/*
 * Build Proof Request for provided template and send it to provided connection
 * */
export const sendProofRequest = async (
  agent: Agent,
  template: ProofRequestTemplate,
  connectionId: string,
  customPredicateValues?: Record<string, Record<string, number>>
): Promise<SendProofRequestResult | undefined> => {
  const proofFormats = buildProofRequestDataForTemplate(template, customPredicateValues)
  if (!proofFormats) {
    return undefined
  }
  const proofRecord = await agent.proofs.requestProof({
    protocolVersion,
    connectionId,
    proofFormats,
  })
  return {
    proofRecord,
  }
}

/*
 * Check if the Proof Request template contains at least one predicate
 * */
export const hasPredicates = (record: ProofRequestTemplate): boolean => {
  if (record.payload.type === ProofRequestType.Indy) {
    return record.payload.data.some((d) => d.requestedPredicates && d.requestedPredicates?.length > 0)
  }
  if (record.payload.type === ProofRequestType.DIF) {
    return false
  }
  return false
}

/*
 * Check if the Proof Request template contains parameterizable predicates
 * */
export const isParameterizable = (record: ProofRequestTemplate): boolean => {
  if (record.payload.type === ProofRequestType.Indy) {
    return record.payload.data.some((d) => d.requestedPredicates?.some((predicate) => predicate.parameterizable))
  }
  if (record.payload.type === ProofRequestType.DIF) {
    return false
  }
  return false
}
