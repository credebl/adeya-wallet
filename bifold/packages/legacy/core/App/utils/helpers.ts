import {
  Agent,
  ConnectionRecord,
  CredentialExchangeRecord,
  RequestedAttribute,
  RequestedPredicate,
  RetrievedCredentials,
  IndyCredentialInfo,
  IndyProofFormat,
} from '@aries-framework/core'
import {
  FormatDataMessagePayload,
  FormatRetrievedCredentialOptions,
} from '@aries-framework/core/build/modules/proofs/models/ProofServiceOptions'
import { useConnectionById } from '@aries-framework/react-hooks'
import { Buffer } from 'buffer'
import { WalletQuery } from 'indy-sdk-react-native'
import moment from 'moment'
import { ParsedUrl, parseUrl } from 'query-string'
import { ReactNode } from 'react'

import { domain } from '../constants'
import { i18n } from '../localization/index'
import { Attribute, Predicate, ProofCredentialAttributes, ProofCredentialPredicates } from '../types/record'
import { ChildFn } from '../types/tour'

export { parsedCredDefNameFromCredential } from './cred-def'
import { parseCredDefFromId } from './cred-def'

export { parsedCredDefName } from './cred-def'
export { parsedSchema } from './schema'

/**
 * Generates a numerical hash based on a given string
 * @see https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript
 * @param { string } s given string
 * @returns { number } numerical hash value
 */
export const hashCode = (s: string): number => {
  return s.split('').reduce((hash, char) => char.charCodeAt(0) + ((hash << 5) - hash), 0)
}

/**
 * Generates a pseudorandom number between 0 and 1 based on a seed
 * @see https://gist.github.com/tommyettinger/46a874533244883189143505d203312c
 * @see https://stackoverflow.com/questions/521295/seeding-the-random-number-generator-in-javascript
 * @param { number } seed any number
 * @returns { number } pseudorandom number between 0 and 1
 */
const mulberry32 = (seed: number) => {
  let t = (seed += 0x6d2b79f5)
  t = Math.imul(t ^ (t >>> 15), t | 1)
  t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
  return ((t ^ (t >>> 14)) >>> 0) / 4294967296
}

/**
 * Converts a numerical hash into a hexidecimal color string
 * @see https://helderesteves.com/generating-random-colors-js/#Generating_random_dark_colors
 * @param { number } hash numerical hash value (generated by hashCode function above)
 * @returns { string } hexidecimal string eg. #32d3cc
 */
export const hashToRGBA = (hash: number) => {
  let color = '#'
  const colorRangeUpperBound = 256

  // once for r, g, b
  for (let i = 0; i < 3; i++) {
    // append a pseudorandom two-char hexidecimal value from the lower half of the color spectrum (to limit to darker colors)
    color += ('0' + Math.floor((mulberry32(hash + i) * colorRangeUpperBound) / 2).toString(16)).slice(-2)
  }

  return color
}

export function formatTime(time: Date, params?: { long?: boolean; format?: string }): string {
  const getMonthKey = 'MMMM'
  const momentTime = moment(time)
  const monthKey = momentTime.format(getMonthKey)
  const customMonthFormatRe = /M+/
  const long = params?.long
  const format = params?.format
  const shortDateFormatMaskLength = 3

  let formatString = i18n.t('Date.ShortFormat')
  if (format) {
    formatString = format
  } else {
    if (long) {
      formatString = i18n.t('Date.LongFormat')
    }

    // if translation fails
    if (formatString === 'Date.ShortFormat' || formatString === 'Date.LongFormat' || formatString === undefined) {
      formatString = 'MMM D, YYYY'
    }
  }
  const customMonthFormat = formatString?.match(customMonthFormatRe)?.[0]

  let formattedTime = momentTime.format(formatString)

  if (customMonthFormat) {
    let monthReplacement = ''
    const monthReplacementKey = momentTime.format(customMonthFormat)
    if (customMonthFormat.length === shortDateFormatMaskLength) {
      // then we know we're dealing with a short date format: 'MMM'
      monthReplacement = i18n.t(`Date.MonthShort.${monthKey}`)
    } else if (customMonthFormat.length > shortDateFormatMaskLength) {
      // then we know we're working with a long date format: 'MMMM'
      monthReplacement = i18n.t(`Date.MonthLong.${monthKey}`)
    }
    // if translation doesn't work
    if (monthReplacement === `Date.MonthLong.${monthKey}` || monthReplacement === `Date.MonthShort.${monthKey}`) {
      monthReplacement = monthReplacementKey
    }

    if (monthReplacement) {
      formattedTime = formattedTime.replace(monthReplacementKey, monthReplacement)
    }
  }
  return formattedTime
}

/**
 * @deprecated The function should not be used
 */
export function connectionRecordFromId(connectionId?: string): ConnectionRecord | void {
  if (connectionId) {
    return useConnectionById(connectionId)
  }
}

/**
 * @deprecated The function should not be used
 */
export function getConnectionName(connection: ConnectionRecord | void): string | void {
  if (!connection) {
    return
  }
  return connection?.alias || connection?.theirLabel
}

export function getCredentialConnectionLabel(credential?: CredentialExchangeRecord) {
  if (!credential) {
    return ''
  }

  if (credential.connectionId) {
    const connection = useConnectionById(credential.connectionId)
    return connection?.alias || connection?.theirLabel || credential.connectionId
  }

  return 'Unknown Contact'
}

export function getConnectionImageUrl(connectionId: string) {
  const connection = useConnectionById(connectionId)
  if (!connection) {
    return undefined
  }
  return connection.imageUrl ?? undefined
}

export function firstValidCredential(
  fields: RequestedAttribute[] | RequestedPredicate[],
  revoked = true
): RequestedAttribute | RequestedPredicate | null {
  if (!fields.length) {
    return null
  }

  let first = null
  const firstNonRevoked = fields.filter((field) => !field.revoked)[0]
  if (firstNonRevoked) {
    first = firstNonRevoked
  } else if (fields.length && revoked) {
    first = fields[0]
  }

  if (!first?.credentialInfo) {
    return null
  }

  return first
}

export const getOobDeepLink = async (url: string, agent: Agent | undefined): Promise<any> => {
  const queryParams = parseUrl(url).query
  const b64Message = queryParams['d_m'] ?? queryParams['c_i']
  const rawmessage = Buffer.from(b64Message as string, 'base64').toString()
  const message = JSON.parse(rawmessage)
  await agent?.receiveMessage(message)
  return message
}

/**
 * A sorting function for the Array `sort()` function
 * @param a First retrieved credential
 * @param b Second retrieved credential
 */
export const credentialSortFn = (a: any, b: any) => {
  if (a.revoked && !b.revoked) {
    return 1
  } else if (!a.revoked && b.revoked) {
    return -1
  } else {
    return b.timestamp - a.timestamp
  }
}

const credNameFromRestriction = (queries?: WalletQuery[]): string => {
  let schema_name = ''
  let cred_def_id = ''
  let schema_id = ''
  queries?.forEach((query) => {
    schema_name = (query?.schema_name as string) ?? schema_name
    cred_def_id = (query?.cred_def_id as string) ?? cred_def_id
    schema_id = (query?.schema_id as string) ?? schema_id
  })
  if (schema_name && (schema_name.toLowerCase() !== 'default' || schema_name.toLowerCase() !== 'credential')) {
    return schema_name
  } else {
    return parseCredDefFromId(cred_def_id, schema_id)
  }
}

export const processProofAttributes = (
  request?: FormatDataMessagePayload<[IndyProofFormat], 'request'> | undefined,
  credentials?: FormatRetrievedCredentialOptions<[IndyProofFormat]>,
  credentialRecords?: CredentialExchangeRecord[]
): { [key: string]: ProofCredentialAttributes } => {
  const processedAttributes = {} as { [key: string]: ProofCredentialAttributes }
  if (!(request?.indy?.requested_attributes && credentials?.proofFormats?.indy?.requestedAttributes)) {
    return {}
  }

  const requestedProofAttributes = request.indy.requested_attributes
  const retrievedCredentialAttributes = credentials.proofFormats.indy.requestedAttributes

  for (const key of Object.keys(retrievedCredentialAttributes)) {
    // The shift operation modifies the original input array, therefore make a copy
    const credential = [...(retrievedCredentialAttributes[key] ?? [])].sort(credentialSortFn).shift()
    const credNameRestriction = credNameFromRestriction(requestedProofAttributes[key]?.restrictions)

    let credName = credNameRestriction ?? key
    if (credential?.credentialInfo?.credentialDefinitionId || credential?.credentialInfo?.schemaId) {
      credName = parseCredDefFromId(
        credential?.credentialInfo?.credentialDefinitionId,
        credential?.credentialInfo?.schemaId
      )
    }
    let revoked = false
    let credExchangeRecord = undefined
    if (credential) {
      credExchangeRecord = credentialRecords?.filter(
        (record) => record.credentials[0]?.credentialRecordId === credential.credentialId
      )[0]
      revoked = credExchangeRecord?.revocationNotification !== undefined
    }
    const { name, names } = requestedProofAttributes[key]

    for (const attributeName of [...(names ?? (name && [name]) ?? [])]) {
      if (!processedAttributes[credName]) {
        // init processedAttributes object
        processedAttributes[credName] = {
          credExchangeRecord,
          schemaId: credential?.credentialInfo?.schemaId,
          credDefId: credential?.credentialInfo?.credentialDefinitionId,
          credName,
          attributes: [],
        }
      }

      let attributeValue = '' //(credentialInfo as IndyCredentialInfo).attributes[attributeName]
      if (credential) {
        attributeValue = (credential.credentialInfo as IndyCredentialInfo).attributes[attributeName]
      }
      processedAttributes[credName].attributes?.push(
        new Attribute({
          revoked,
          name: attributeName,
          value: attributeValue,
        })
      )
    }
  }
  return processedAttributes
}

export const processProofPredicates = (
  request?: FormatDataMessagePayload<[IndyProofFormat], 'request'> | undefined,
  credentials?: FormatRetrievedCredentialOptions<[IndyProofFormat]>,
  credentialRecords?: CredentialExchangeRecord[]
): { [key: string]: ProofCredentialPredicates } => {
  const processedPredicates = {} as { [key: string]: ProofCredentialPredicates }

  if (!(request?.indy?.requested_predicates && credentials?.proofFormats?.indy?.requestedPredicates)) {
    return {}
  }

  const requestedProofPredicates = request.indy.requested_predicates
  const retrievedCredentialPredicates = credentials.proofFormats.indy.requestedPredicates

  for (const key of Object.keys(requestedProofPredicates)) {
    // The shift operation modifies the original input array, therefore make a copy
    const credential = [...(retrievedCredentialPredicates[key] ?? [])].sort(credentialSortFn).shift()
    let credExchangeRecord = undefined
    if (credential) {
      credExchangeRecord = credentialRecords?.filter(
        (record) => record.credentials[0]?.credentialRecordId === credential.credentialId
      )[0]
    }
    const { credentialId, credentialDefinitionId, schemaId } = { ...credential, ...credential?.credentialInfo }
    const revoked =
      credentialRecords?.filter((record) => record.credentials[0]?.credentialRecordId === credentialId)[0]
        ?.revocationNotification !== undefined
    const { name, p_type: pType, p_value: pValue } = requestedProofPredicates[key]

    const credNameRestriction = credNameFromRestriction(requestedProofPredicates[key]?.restrictions)

    let credName = credNameRestriction ?? key
    if (credential?.credentialInfo?.credentialDefinitionId || credential?.credentialInfo?.schemaId) {
      credName = parseCredDefFromId(
        credential?.credentialInfo?.credentialDefinitionId,
        credential?.credentialInfo?.schemaId
      )
    }

    if (!processedPredicates[credName]) {
      processedPredicates[credName] = {
        credExchangeRecord,
        schemaId,
        credDefId: credentialDefinitionId,
        credName: credName,
        predicates: [],
      }
    }

    processedPredicates[credName].predicates?.push(
      new Predicate({
        credentialId,
        name,
        revoked,
        pValue,
        pType,
      })
    )
  }
  return processedPredicates
}

export const mergeAttributesAndPredicates = (
  attributes: { [key: string]: ProofCredentialAttributes },
  predicates: { [key: string]: ProofCredentialPredicates }
) => {
  const merged = { ...attributes }
  for (const [key, predicate] of Object.entries(predicates)) {
    const existingEntry = merged[key]
    if (existingEntry) {
      merged[key] = { ...existingEntry, ...predicate }
    } else {
      merged[key] = predicate
    }
  }
  return merged
}

/**
 * @deprecated The function should not be used
 */
export const sortCredentialsForAutoSelect = (credentials: RetrievedCredentials): RetrievedCredentials => {
  const requestedAttributes = Object.values(credentials?.requestedAttributes).pop()
  const requestedPredicates = Object.values(credentials?.requestedPredicates).pop()
  const sortFn = (a: any, b: any) => {
    if (a.revoked && !b.revoked) {
      return 1
    } else if (!a.revoked && b.revoked) {
      return -1
    } else {
      return b.timestamp - a.timestamp
    }
  }

  requestedAttributes && requestedAttributes.sort(sortFn)
  requestedPredicates && requestedPredicates.sort(sortFn)

  return credentials
}

/**
 *
 * @param url a redirection URL to retrieve a payload for an invite
 * @param agent an Agent instance
 * @returns payload from following the redirection
 */
export const receiveMessageFromUrlRedirect = async (url: string, agent: Agent | undefined) => {
  const res = await fetch(url, {
    method: 'GET',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  })
  const message = await res.json()
  await agent?.receiveMessage(message)
  return message
}

/**
 *
 * @param url a redirection URL to retrieve a payload for an invite
 * @param agent an Agent instance
 * @returns payload from following the redirection
 */
export const receiveMessageFromDeepLink = async (url: string, agent: Agent | undefined) => {
  const res = await fetch(url, {
    method: 'GET',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
  })
  const message = await res.json()
  await agent?.receiveMessage(message)
  return message
}

/**
 *
 * @param uri a URI containing a base64 encoded connection invite in the query parameter
 * @param agent an Agent instance
 * @returns a connection record from parsing and receiving the invitation
 */
export const connectFromInvitation = async (uri: string, agent: Agent | undefined) => {
  const invitation = await agent?.oob.parseInvitation(uri)

  if (!invitation) {
    throw new Error('Could not parse invitation from URL')
  }

  const record = await agent?.oob.receiveInvitation(invitation)
  const connectionRecord = record?.connectionRecord
  if (!connectionRecord?.id) {
    throw new Error('Connection does not have an ID')
  }

  return connectionRecord
}

/**
 * Create a new connection invitation
 *
 * @param agent an Agent instance
 * @returns a connection record
 */
export const createConnectionInvitation = async (agent: Agent | undefined) => {
  const record = await agent?.oob.createInvitation()
  if (!record) {
    throw new Error('Could not create new invitation')
  }
  const invitationUrl = record.outOfBandInvitation.toUrl({ domain })
  return {
    record,
    invitation: record.outOfBandInvitation,
    invitationUrl,
  }
}

/**
 * Parse URL from provided string
 * @param urlString string to parse
 * @returns ParsedUur object if success or undefined
 */
export const getUrl = (urlString: string): ParsedUrl | undefined => {
  try {
    return parseUrl(urlString)
  } catch (e) {
    return undefined
  }
}

/**
 * Parse JSON from provided string
 * @param jsonString string to parse
 * @returns JSON object if success or undefined
 */
export const getJson = (jsonString: string): Record<string, unknown> | undefined => {
  try {
    return JSON.parse(jsonString)
  } catch (e) {
    return undefined
  }
}

/**
 * Typeguard to check if any React children is represented as a function
 * instead of a Node. I,e., when it's a {@link ChildFn}.
 *
 * @param children any React children
 * @returns true if the children is a function, false otherwise
 */
export function isChildFunction<T>(children: ReactNode | ChildFn<T>): children is ChildFn<T> {
  return typeof children === 'function'
}
