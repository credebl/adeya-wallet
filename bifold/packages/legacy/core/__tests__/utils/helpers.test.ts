import fs from 'fs'
import path from 'path'

import { credentialSortFn } from '../../App/utils/helpers'

const proofCredentialPath = path.join(__dirname, '../fixtures/proof-credential.json')
const credentials = JSON.parse(fs.readFileSync(proofCredentialPath, 'utf8'))

describe('Helpers', () => {
  test('Sorts retrieved credentials by revocation', async () => {
    const key = '0_surname_uuid'
    const { requestedAttributes } = credentials
    const sortedCreds = requestedAttributes[key].sort(credentialSortFn)

    expect(sortedCreds).toMatchSnapshot()
  })
})
