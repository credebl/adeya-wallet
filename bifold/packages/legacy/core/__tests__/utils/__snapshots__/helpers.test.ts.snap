// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers Sorts retrieved credentials by revocation 1`] = `
Array [
  Object {
    "credentialId": "4e91a381-7ed6-4ea7-8011-39a11cbd0c18",
    "credentialInfo": Object {
      "attrs": Object {
        "Given Name": "Francis",
        "Member Status": "Practising",
        "Member Status Code": "PRAC",
        "PPID": "MC12349",
        "Surname": "Goodrangsy",
      },
      "cred_def_id": "AuJrigKQGRLJajKAebTgWu:3:CL:209526:default",
      "cred_rev_id": "11",
      "referent": "4e91a381-7ed6-4ea7-8011-39a11cbd0c18",
      "rev_reg_id": "AuJrigKQGRLJajKAebTgWu:4:AuJrigKQGRLJajKAebTgWu:3:CL:209526:default:CL_ACCUM:7aa88d5c-e1e8-488c-a0a5-63aee1b0ae66",
      "schema_id": "AuJrigKQGRLJajKAebTgWu:2:Member Card:1.5.1",
    },
    "revealed": true,
    "revoked": false,
    "timestamp": 1654889164,
  },
  Object {
    "credentialId": "d1092422-08d0-448f-8ea0-fd7f33018277",
    "credentialInfo": Object {
      "attrs": Object {
        "Given Name": "Francis",
        "Member Status": "Practising",
        "Member Status Code": "PRAC",
        "PPID": "MC12349",
        "Surname": "Goodrangsy",
      },
      "cred_def_id": "AuJrigKQGRLJajKAebTgWu:3:CL:209526:default",
      "cred_rev_id": "9",
      "referent": "d1092422-08d0-448f-8ea0-fd7f33018277",
      "rev_reg_id": "AuJrigKQGRLJajKAebTgWu:4:AuJrigKQGRLJajKAebTgWu:3:CL:209526:default:CL_ACCUM:7aa88d5c-e1e8-488c-a0a5-63aee1b0ae66",
      "schema_id": "AuJrigKQGRLJajKAebTgWu:2:Member Card:1.5.1",
    },
    "revealed": true,
    "revoked": true,
    "timestamp": 1654889165,
  },
  Object {
    "credentialId": "8fb6c834-76e8-46ac-9c75-6a5e984e6c6a",
    "credentialInfo": Object {
      "attrs": Object {
        "Given Name": "Francis",
        "Member Status": "Practising",
        "Member Status Code": "PRAC",
        "PPID": "MC12349",
        "Surname": "Goodrangsy",
      },
      "cred_def_id": "AuJrigKQGRLJajKAebTgWu:3:CL:209526:default",
      "cred_rev_id": "10",
      "referent": "8fb6c834-76e8-46ac-9c75-6a5e984e6c6a",
      "rev_reg_id": "AuJrigKQGRLJajKAebTgWu:4:AuJrigKQGRLJajKAebTgWu:3:CL:209526:default:CL_ACCUM:7aa88d5c-e1e8-488c-a0a5-63aee1b0ae66",
      "schema_id": "AuJrigKQGRLJajKAebTgWu:2:Member Card:1.5.1",
    },
    "revealed": true,
    "revoked": true,
    "timestamp": 1654889163,
  },
]
`;
